### PR TITLE
model-conversion : add support for SentenceTransformers

### PR DESCRIPTION
**Pull request by danbev**
This commit adds support for models that use SentenceTransformer layers.

The motivation for this is that if converted model includes any of the numbered layers specified in the original models repository then these changes enable these models to be used and verified. Currently the model-conversion only support the base model output without any of the additional transformation layers.

Usage:
Convert the model that also includes the SentenceTransformer layers:
```console
(venv) $ export EMBEDDING_MODEL_PATH="~/google/embeddinggemma-300M"
(venv) make embedding-convert-model-st
```

Verify the produced embeddings from the converted model against the original model embeddings:
```console
(venv) make embedding-verify-logits-st
```

The original model can be run using SentenceTransformer:
```console
(venv) make embedding-run-original-model-st
```

Run the converted model using "SentenceTransformer" layers whic enables pooling and normalization:
```console
(venv) make embedding-run-converted-model-st
```

### CODE REVIEW

File: examples/model-conversion/logits.cpp [hunk 1]
```diff
@@ -164,43 +186,49 @@ int main(int argc, char ** argv) {
...
 
             printf(" ... ");
 
             // Print last 3 values
             for (int i = n_embd - 3; i < n_embd; i++) {
                 if (i >= 0) {
-                    printf("%9.6f ", logits[j * n_embd + i]);
+                    printf("%9.6f ", embeddings[j * n_embd + i]);
                 }
             }
```
Comment (tommarques56) [inline examples/model-conversion/logits.cpp:200]: **Issue:** The embd_norm parameter is not validated to ensure it's within acceptable range before being used

**Suggestion:** Add validation to ensure embd_norm is a valid value (>= -1) before using it in normalization

**Code to review:**
```diff
printf("Embeddings size: %d\n", n_embeddings);
```

**Suggested fix:**
```
if (embd_norm < -1) { fprintf(stderr, "Invalid embedding normalization value: %d\n", embd_norm); return 1; }
```

*Automated review by PR Code Review Service*

File: examples/model-conversion/logits.cpp [hunk 2]
```diff
@@ -41,9 +49,13 @@ int main(int argc, char ** argv) {
                     return 1;
                 }
             } else if (strcmp(argv[i], "-embd-mode") == 0) {
+                embedding_mode = true;
+            } else if (strcmp(argv[i], "-pooling") == 0) {
+                pooling_enabled = true;
+            } else if (strcmp(argv[i], "-embd-norm") == 0) {
                 if (i + 1 < argc) {
```
Comment (tommarques56) [inline examples/model-conversion/logits.cpp:56]: **Issue:** The code parses -embd-norm flag without validating the range of embd_norm values, which may cause unexpected behavior for invalid normalization types.

**Suggestion:** Add validation to ensure embd_norm is within acceptable range (-1 to 2) or is a positive integer for p-norm.

**Code to review:**
```diff
                        embd_norm = std::stoi(argv[++i]);
```

**Suggested fix:**
```
if (embd_norm < -1) { fprintf(stderr, "Invalid embd_norm value: %d\n", embd_norm); return 1; }
```

*Automated review by PR Code Review Service*

File: examples/model-conversion/logits.cpp [hunk 3]
```diff
@@ -8,7 +11,10 @@
 
 static void print_usage(int, char ** argv) {
     printf("\nexample usage:\n");
-    printf("\n    %s -m model.gguf [-ngl n_gpu_layers] -embd-mode [prompt]\n", argv[0]);
+    printf("\n    %s -m model.gguf [-ngl n_gpu_layers] -embd-mode [-pooling] [-embd-norm <norm>] [prompt]\n", argv[0]);
```
Comment (tommarques56) [inline examples/model-conversion/logits.cpp:14]: **Issue:** Unnecessary code duplication in parameter parsing

**Suggestion:** Refactor duplicated code into a helper function to reduce redundancy

**Code to review:**
```diff
printf("%s: error: unable to load model\n" , __func__);
```

*Automated review by PR Code Review Service*

File: examples/model-conversion/logits.cpp [hunk 4]
```diff
@@ -164,43 +186,49 @@ int main(int argc, char ** argv) {
...
     std::filesystem::create_directory("data");
 
-    // Save logits to binary file
+    // Save data to binary file
     char bin_filename[512];
     snprintf(bin_filename, sizeof(bin_filename), "data/llamacpp-%s%s.bin", model_name, type);
-    printf("Saving logits to %s\n", bin_filename);
+    printf("Saving data to %s\n", bin_filename);
 
     FILE * f = fopen(bin_filename, "wb");
```
Comment (tommarques56) [inline examples/model-conversion/logits.cpp:226]: **Issue:** Missing error checking for file operations when writing binary data

**Suggestion:** Add checks for fwrite return value to ensure data was written successfully

**Code to review:**
```diff
fwrite(data_ptr, sizeof(float), data_size, f);
```

*Automated review by PR Code Review Service*

File: examples/model-conversion/logits.cpp [hunk 5]
```diff
@@ -211,27 +239,27 @@ int main(int argc, char ** argv) {
```
Comment (tommarques56) [inline examples/model-conversion/logits.cpp:210]: **Issue:** Missing error checking for file operations when writing text data

**Suggestion:** Add checks for fprintf return values to ensure data was written successfully

**Code to review:**
```diff
fprintf(f, "%d: %.6f\n", i, data_ptr[i]);
```

*Automated review by PR Code Review Service*

File: examples/model-conversion/logits.cpp [hunk 6]
```diff
@@ -8,7 +11,10 @@
 
 static void print_usage(int, char ** argv) {
     printf("\nexample usage:\n");
-    printf("\n    %s -m model.gguf [-ngl n_gpu_layers] -embd-mode [prompt]\n", argv[0]);
+    printf("\n    %s -m model.gguf [-ngl n_gpu_layers] -embd-mode [-pooling] [-embd-norm <norm>] [prompt]\n", argv[0]);
+    printf("\n");
```
Comment (tommarques56) [inline examples/model-conversion/logits.cpp:15]: **Issue:** The usage message for command-line arguments is updated but doesn't fully reflect the new options.

**Suggestion:** Ensure usage message includes all new command-line flags and their descriptions.

**Code to review:**
```diff
    printf("  -embd-norm: normalization type for pooled embeddings (default: 2)\n");
```

**Suggested fix:**
```
printf("  -embd-norm: normalization type for pooled embeddings (default: 2)\n");
```

*Automated review by PR Code Review Service*

File: examples/model-conversion/scripts/embedding/run-converted-model.sh [hunk 7]
```diff
@@ -47,4 +52,8 @@ echo $CONVERTED_MODEL
 
 cmake --build ../../build --target llama-logits -j8
 # TODO: update logits.cpp to accept a --file/-f option for the prompt
-../../build/bin/llama-logits -m "$CONVERTED_MODEL" -embd-mode "$PROMPT"
```
Comment (tommarques56) [inline examples/model-conversion/scripts/embedding/run-converted-model.sh:50]: **Issue:** The script does not validate that the model file exists before attempting to use it, which could result in a cryptic error message.

**Suggestion:** Add a check to ensure the model file exists before proceeding.

**Code to review:**
```diff
if [ -n "$USE_POOLING" ]; then
```

**Suggested fix:**
```
if [ ! -f "$CONVERTED_MODEL" ]; then echo "Error: Model file '$CONVERTED_MODEL' not found" >&2; exit 1; fi
```

*Automated review by PR Code Review Service*

File: examples/model-conversion/scripts/embedding/run-converted-model.sh [hunk 8]
```diff
@@ -5,13 +5,18 @@ set -e
 # Parse command line arguments
 CONVERTED_MODEL=""
 PROMPTS_FILE=""
+USE_POOLING=""
```
Comment (tommarques56) [inline examples/model-conversion/scripts/embedding/run-converted-model.sh:8]: **Issue:** Variable initialization for USE_POOLING is a good practice but should be done in a way that's consistent with other variable declarations.

**Suggestion:** Initialize USE_POOLING with empty string at the same time as other variables for consistency.

**Code to review:**
```diff
+USE_POOLING=""
```

**Suggested fix:**
```
CONVERTED_MODEL=""
PROMPTS_FILE=""
USE_POOLING=""
```

*Automated review by PR Code Review Service*

File: requirements/requirements-all.txt [hunk 9]
```diff
@@ -14,3 +14,5 @@
 -r ./requirements-tool_bench.txt
 
 -r ./requirements-gguf_editor_gui.txt
+
+-r ../examples/model-conversion/requirements.txt
```
Comment (CISC) [inline requirements/requirements-all.txt:18]: This should have triggered `Python check requirements.txt` and `Python Type-Check` actions.

As this might have implications for those I'd like them to run first, unfortunately they are not manually triggerable...

Comment (danbev) [reply requirements/requirements-all.txt:18]: I've rebased this now and the checks should run now.

**Review by CISC — APPROVED** [review]
